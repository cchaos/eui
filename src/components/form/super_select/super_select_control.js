import React, { Fragment } from 'react';
import { renderToString } from 'react-dom/server';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import {
  EuiFormControlLayout,
} from '../form_control_layout';

export const EuiSuperSelectControl = ({
  className,
  options,
  id,
  name,
  inputRef,
  fullWidth,
  isLoading,
  hasNoInitialSelection,
  defaultValue,
  compressed,
  value,
  ...rest
}) => {
  const classes = classNames(
    'euiSuperSelectControl',
    {
      'euiSuperSelectControl--fullWidth': fullWidth,
      'euiSuperSelectControl--compressed': compressed,
      'euiSuperSelectControl-isLoading': isLoading,
    },
    className
  );

  let emptyOptionNode;
  if (hasNoInitialSelection) {
    emptyOptionNode = (
      <option value="" disabled hidden style={{ display: 'none' }}>&nbsp;</option>
    );
  }

  // React HTML input can not have both value and defaultValue properties.
  // https://reactjs.org/docs/uncontrolled-components.html#default-values
  let selectDefaultValue;
  if (!value) {
    selectDefaultValue = defaultValue || '';
  }

  let selectedValue;
  if (value) {
    const selectedOption = options.find(option => option.value === value);
    selectedValue = selectedOption.inputDisplay;
  }

  const icon = {
    type: 'arrowDown',
    side: 'right',
  };

  // In the case the inputDisplay is a react element, make it a string, then strip the HTML
  // This is used to read only the text for a screenreader.
  const selectedValueAsString = renderToString(selectedValue).replace(/<\/?[^>]+(>|$)/g, '');

  return (
    <Fragment>
      <select
        id={id}
        name={name}
        className="euiSuperSelectControl__hiddenField"
        ref={inputRef}
        defaultValue={selectDefaultValue}
        value={value}
        aria-hidden="true"
        {...rest}
      >
        {emptyOptionNode}
        {options.map((option, index) => {
          const {
            text, // eslint-disable-line no-unused-vars
            value,
          } = option;
          return <option value={value} key={index} />;
        })}
      </select>

      <EuiFormControlLayout
        icon={icon}
        fullWidth={fullWidth}
        isLoading={isLoading}
        compressed={compressed}
      >

        <button
          role="option"
          type="button"
          className={classes}
          aria-haspopup="true"
          aria-label={`Select an option: ${selectedValueAsString} selected`}
          {...rest}
        >
          {selectedValue}
        </button>

      </EuiFormControlLayout>
    </Fragment>
  );
};

EuiSuperSelectControl.propTypes = {
  name: PropTypes.string,
  id: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.node.isRequired,
    inputDisplay: PropTypes.node.isRequired
  })).isRequired,
  isInvalid: PropTypes.bool,
  fullWidth: PropTypes.bool,
  isLoading: PropTypes.bool,

  /**
   * Simulates no selection by creating an empty, selected, hidden first option
   */
  hasNoInitialSelection: PropTypes.bool,
  inputRef: PropTypes.func,
  /**
   * when `true` creates a shorter height input
   */
  compressed: PropTypes.bool,
};

EuiSuperSelectControl.defaultProps = {
  options: [],
  fullWidth: false,
  isLoading: false,
  hasNoInitialSelection: false,
  compressed: false,
};
