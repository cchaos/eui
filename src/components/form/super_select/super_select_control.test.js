import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test';

import { EuiSuperSelectControl } from './super_select_control';

describe('EuiSuperSelectControl', () => {
  test('is rendered', () => {
    const component = render(
      <EuiSuperSelectControl {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
