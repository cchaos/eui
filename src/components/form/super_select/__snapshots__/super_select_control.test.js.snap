// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSuperSelectControl is rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    aria-label="aria-label"
    class="euiSuperSelectControl__hiddenField"
    data-test-subj="test subject string"
  />,
  <div
    class="euiFormControlLayout"
  >
    <button
      aria-label="aria-label"
      class="euiSuperSelectControl testClass1 testClass2"
      data-test-subj="test subject string"
      role="option"
      type="button"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;

exports[`EuiSuperSelectControl props compressed is rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    class="euiSuperSelectControl__hiddenField"
  />,
  <div
    class="euiFormControlLayout euiFormControlLayout--compressed"
  >
    <button
      class="euiSuperSelectControl euiSuperSelectControl--compressed"
      role="option"
      type="button"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;

exports[`EuiSuperSelectControl props disabled options are rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    class="euiSuperSelectControl__hiddenField"
  >
    <option
      value="1"
    />
    <option
      value="2"
    />
  </select>,
  <div
    class="euiFormControlLayout"
  >
    <button
      class="euiSuperSelectControl"
      role="option"
      type="button"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;

exports[`EuiSuperSelectControl props fullWidth is rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    class="euiSuperSelectControl__hiddenField"
  />,
  <div
    class="euiFormControlLayout euiFormControlLayout--fullWidth"
  >
    <button
      class="euiSuperSelectControl euiSuperSelectControl--fullWidth"
      role="option"
      type="button"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;

exports[`EuiSuperSelectControl props isLoading is rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    class="euiSuperSelectControl__hiddenField"
  />,
  <div
    class="euiFormControlLayout"
  >
    <button
      class="euiSuperSelectControl euiSuperSelectControl-isLoading"
      role="option"
      type="button"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <div
        class="euiLoadingSpinner euiLoadingSpinner--medium"
      />
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;

exports[`EuiSuperSelectControl props options are rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    class="euiSuperSelectControl__hiddenField"
  >
    <option
      value="1"
    />
    <option
      value="2"
    />
  </select>,
  <div
    class="euiFormControlLayout"
  >
    <button
      class="euiSuperSelectControl"
      role="option"
      type="button"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;

exports[`EuiSuperSelectControl props value option is rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    class="euiSuperSelectControl__hiddenField"
  >
    <option
      selected=""
      value="1"
    />
    <option
      value="2"
    />
  </select>,
  <div
    class="euiFormControlLayout"
  >
    <button
      class="euiSuperSelectControl"
      role="option"
      type="button"
    >
      Option #1
    </button>
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;
